fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/keyspace/KeySpacePathImpl.java

```diff
@@ -276,27 +276,22 @@ public boolean equals(Object obj) {
         }
         KeySpacePath that = (KeySpacePath) obj;
 
-        // Check that the KeySpaceDirectories of the two paths are "equal enough".
-        // Even this is probably overkill since the isCompatible check in KeySpaceDirectory
-        // will keep us from doing anything too bad. We could move this check into KeySpaceDirectory
-        // but comparing two directories by value would necessitate traversing the entire directory
-        // tree, so instead we will use a narrower definition of equality here.
-        boolean directoriesEqual = Objects.equals(this.getDirectory().getKeyType(), that.getDirectory().getKeyType()) &&
-                                   Objects.equals(this.getDirectory().getName(), that.getDirectory().getName()) &&
-                                   Objects.equals(this.getDirectory().getValue(), that.getDirectory().getValue());
+        // We don't care whether the two objects exist in the same hierarchy, we validate the relevant bits by comparing
+        // parents.
+        boolean directoriesEqual = this.getDirectory().equalsIgnoringHierarchy(that.getDirectory());
 
+        // the values might be byte[]
         return directoriesEqual &&
-               Objects.equals(this.getValue(), that.getValue()) &&
-               Objects.equals(this.getParent(), that.getParent());
+                KeySpaceDirectory.areEqual(this.getValue(), that.getValue()) &&
+                Objects.equals(this.getParent(), that.getParent());
     }
 
     @Override
     public int hashCode() {
         return Objects.hash(
                 getDirectory().getKeyType(),
                 getDirectory().getName(),
-                getDirectory().getValue(),
-                getValue(),
+                KeySpaceDirectory.valueHashCode(getValue()),
```

Comment:
This would be the same as `getDirectory().getValue()`, right?

Reply:
Only if the directory is a constant. If the `getDirectory().getValue()` is `ANY_VALUE`, this would be any value of that type.